libs/providers/langchain-mcp-adapters/src/interceptor.ts

```diff
@@ -0,0 +1,168 @@
+import { z } from "zod/v3";
+import {
+  LoggingMessageNotificationSchema,
+  ProgressSchema,
+  EmbeddedResourceSchema,
+} from "@modelcontextprotocol/sdk/types.js";
+import type { ContentBlock } from "@langchain/core/messages";
+import type { RunnableConfig } from "@langchain/core/runnables";
+
+export { RunnableConfig };
+
+const toolCallRequestSchema = z.object({
+  serverName: z.string(),
+  name: z.string(),
+  args: z.unknown(),
+});
+export type ToolCallRequest = z.output<typeof toolCallRequestSchema>;
+
+const toolResultSchema = z.tuple([
+  z.custom<string | (ContentBlock | ContentBlock.Data.DataContentBlock)[]>(),
+  z.array(
+    z.union([
+      EmbeddedResourceSchema,
+      z.custom<ContentBlock.Data.DataContentBlock>(),
+    ])
+  ),
+]);
+export type ToolResult = z.output<typeof toolResultSchema>;
+
+const toolCallResultSchema = z.object({
+  ...toolCallRequestSchema.shape,
+  result: toolResultSchema,
+});
+export type ToolCallResult = z.output<typeof toolCallResultSchema>;
+
+const toolCallModificationSchema = z
+  .object({
+    header: z.record(z.string()),
+    args: z.unknown(),
+  })
+  .partial();
+export type ToolCallModification = z.output<typeof toolCallModificationSchema>;
+
+export const interceptorSchema = z.object({
+  /**
+   * Called before a tool call is made.
+   * Allows you to modify the tool call arguments or return a different tool call.
+   *
+   * @param toolCallRequest - The tool call request
+   * @param toolCallRequest.name - The tool name
+   * @param toolCallRequest.args - The tool call arguments
+   * @param toolCallRequest.serverName - The server name
+   * @param config - The runnable config
+   * @returns The tool call modification
+   *
+   * @example
+   * ```ts
+   * const interceptor = {
+   *   beforeToolCall: (toolCallRequest) => {
+   *     return {
+   *       args: {
+   *         ...toolCallRequest.args,
+   *         custom: "Custom Value"
+   *       },
+   *       header: { "X-Custom-Header": "Custom Value" }
+   *     };
+   *   },
+   * };
+   * ```
+   */
+  beforeToolCall: z
+    .function()
+    .args(toolCallRequestSchema, z.custom<RunnableConfig>())
```

Comment:
How could we support context when running within langgraph? (i.e., langgraph state + runtime context)

Reply:
By passing in `RunnableConfig` as second arg we provide runtime context information. Can we access state within tool calls?

Reply:
> * State (not captured by runnable config)

Is there a way I can access state?